jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java

```diff
@@ -356,28 +362,77 @@ public ContextHandler createContextHandler(final App app) throws Exception
 
-            //add in environment-specific properties
+            //Add in environment-specific properties:
+            // allow multiple eeXX[-zzz].properties files, ordered lexically
+            // allow each to contain jetty.deploy.environmentXml[.zzzz] properties
+            // accumulate all properties for substitution purposes
+            // order all jetty.deploy.environmentXml[.zzzz] properties lexically
+            // apply the context xml files named by the ordered jetty.deploy.environmentXml[.zzzz] properties
             String env = app.getEnvironmentName() == null ? "" : app.getEnvironmentName();
-            Path envProperties = app.getPath().getParent().resolve(env + ".properties");
-            if (Files.exists(envProperties))
+
+            List<Path> envPropertyFiles = new ArrayList<>();
+            Path parent = app.getPath().getParent();
+
+            //Get all environment specific properties files for this environment,
+            //order them according to the lexical ordering of the filenames
+            try (Stream<Path> paths = Files.list(parent))
             {
-                try (InputStream stream = Files.newInputStream(envProperties))
-                {
-                    Properties p = new Properties();
-                    p.load(stream);
-                    p.stringPropertyNames().forEach(k -> properties.put(k, p.getProperty(k)));
-                }
+                envPropertyFiles = paths.filter(Files::isRegularFile)
+                    .map(p -> parent.relativize(p))
+                    .filter(p ->
+                    {
+                        String name = p.getName(0).toString();
+                        if (!name.endsWith(".properties"))
+                            return false;
+                        if (!name.startsWith(env))
+                            return false;
+                        return true;
+                    }).sorted().collect(Collectors.toList());
+            }
 
-                String str = properties.get(Deployable.ENVIRONMENT_XML);
-                if (!StringUtil.isEmpty(str))
-                {
-                    Path envXmlPath = Paths.get(str);
-                    if (!envXmlPath.isAbsolute())
-                        envXmlPath = getMonitoredDirResource().getPath().getParent().resolve(envXmlPath);
+            if (LOG.isDebugEnabled())
+                LOG.debug("Environment property files {}", envPropertyFiles);
 
-                    context = applyXml(context, envXmlPath, env, properties);
+            Map<String, Path> envXmlFilenameMap = new HashMap<>();
+            for (Path file : envPropertyFiles)
+            {
+                Path resolvedFile = parent.resolve(file);
+                if (Files.exists(resolvedFile))
+                {
+                    Properties tmp = new Properties();
+                    try (InputStream stream = Files.newInputStream(resolvedFile))
```

Comment:
Aren't these meant to be merged into the single properties file and then that is checed properties with the prefix?

Reply:
I don't understand your question, if you look at lines 406 and 407 the properties are all added into the general property substitution pool:
 
``` java
                    try (InputStream stream = Files.newInputStream(resolvedFile))
                    {
                        tmp.load(stream);
                        //put each property into our substitution pool
                        tmp.stringPropertyNames().forEach(k -> properties.put(k, tmp.getProperty(k)));
```